src/Storage/CookieLocaleStorage.php

```diff
@@ -0,0 +1,58 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Yii\Middleware\Storage;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Yiisoft\Cookies\Cookie;
+
+final class CookieLocaleStorage implements LocaleStorageWithHttpFlowInterface
+{
+    private ?ServerRequestInterface $request = null;
+    private ?ResponseInterface $response = null;
+
+    public function __construct(
+        private Cookie $cookie,
+    ) {
+    }
+
+    public function getName(): string
+    {
+        return 'cookie';
+    }
+
+    public function set(string $value): void
+    {
+        $this->cookie = $this->cookie->withValue($value);
+
+        if ($this->response !== null) {
```

Comment:
Looks like performance degradation. What to you think about "flush" stage instead?